thread.c

```diff
@@ -857,6 +861,12 @@ thread_start_func_2(rb_thread_t *th, VALUE *stack_start)
         th->value = Qnil;
```

Comment:
If `Scheduler#unblock` raises, it back to this line and `th->value = Qnil` is assigned. They should be `th->value = Qnil if th->value == Qundef`. Or catch the exception for each `#unblock` call.

Reply:
I tried this out.

```ruby
require_relative 'test/fiber/scheduler'

class RaiseUnblockScheduler < Scheduler
  def unblock(blocker, fiber)
    raise "Broken unblock!"
  end
end

promise = Thread.new do
	sleep 1
	"Hello World"
end

pp promise

thread = Thread.new do
	scheduler = BrokenUnblockScheduler.new

	Fiber.set_scheduler scheduler

	Fiber.schedule do
		begin
			promise.join
		ensure
			$stderr.puts "Fiber promise.join: #{$!}"
		end
	end
end

begin
	thread.join
rescue
	$stderr.puts "Main promise.join: #{$!}"
end

pp thread.value
```

If `unblock` throws an exception, it propagates out of the thread, and the thread value is ignored.

So, even if the value is set, if `unblock` raises, it's considered the thread failed. I think that's okay, since this represents a logical fatal flaw in the program design rather than something that can be recovered.

Essentially, once the exception code is triggered, the `thread.value` will raise that exception, so even if we set it to `Qnil`, I don't think it matters, since it will never be accessible.

The alternative is to some how catch exceptions for `#unblock` call as you say, we would need to print it out, but it wouldn't cause the program to fail. I think this is worse outcome, since it means the program is now in invalid state.

Reply:
OK.